### PR TITLE
[Feat] Add grad_dtype support

### PR DESCRIPTION
**Pull request by 0xWJ**
Introduce support for the `grad_dtype` attribute in parameters, enhancing flexibility in gradient precision management.

### CODE REVIEW

File: nnscaler/runtime/adapter/reducer.py [hunk 1]
```diff
@@ -880,8 +882,9 @@ def add_param(self, param: torch.nn.Parameter):
 
     def _allocate_buffers(self):
         # gradient buffer
+        grad_dtype = get_grad_dtype(self._params[0])
         self._contiguous_grads: torch.Tensor = torch.zeros(
-            (self.buffer_length,), dtype=self._params[0].dtype,
+            (self.buffer_length,), dtype=grad_dtype,
             device=torch.cuda.current_device(), requires_grad=False)
```
[the remark this answers is not in the document]
Comment (0xWJ) [reply nnscaler/runtime/adapter/reducer.py:888]: no fix

File: docs/source/trainer.md [hunk 2]
```diff
@@ -851,11 +851,26 @@ Please note
   as the initial state dict of the model. Default is `None`.
 - `precision`(`Union[str, Dict[_TENSOR_TYPE, _PRECISION_TYPE], None]`):
   The precision of the model. It can be a `str`, which means the same
-  precision for all tensors, or a `Dict[_TENSOR_TYPE, _PRECISION_TYPE]`,
+  precision for all tensors (except `grad`), or a `Dict[_TENSOR_TYPE, _PRECISION_TYPE]`,
   which means the precision for each tensor type. Default is `None`.
-  Currently we support 3 tensor types (`param`, `buffer`, `input`) and
+  Currently we support 4 tensor types (`param`, `buffer`, `input`, `grad`) and
```
Comment (zyeric) [inline docs/source/trainer.md:856]: what is the relationship between these dtypes and optimizer's dtype?

Comment (0xWJ) [reply docs/source/trainer.md:856]: there is no option for optimizer state dtype. For most implementation of optimizers, the state dtype is the same with parameter dtype. So to support grad dtype, the optimizer needs to support mixed precision (like MixedPrecisionAdamW in nnscaler).

**Review by zyeric — APPROVED** [review]

**Review by lynex — APPROVED** [review]
